We don't seem to be updating certificates when updating the connection

### Describe the bug

There's something wrong in the way we upload the certs to the server, I wanted to update the certs and hit the following issue:
- if I update the certs, I get a "wrong cert" error
- if I clean al files in the /tmp/ folder and then I update the certs, I get a "Could not open SSL certificate file /tmp/metabase-secret" error

### To Reproduce

1) Create a Postgres server on GCP. The postgres needs to only accept SSL connections so you'll need the server-ca, the client cert and the client key (don't forget to convert the client key to a DER format)
2) Now add the DB to Metabase
3) Rotate the cert in the DB
4) now update the cert in Metabase, you'll hit an error

### Expected behavior

Certs should be able to be upgraded

### Logs

Caused by: java.io.FileNotFoundException: /tmp/metabase-secret_9099912732785039104.tmp (No such file or directory)
	at java.base/java.io.FileInputStream.open0(Native Method)
	at java.base/java.io.FileInputStream.open(Unknown Source)
	at java.base/java.io.FileInputStream.<init>(Unknown Source)
	at java.base/java.io.FileInputStream.<init>(Unknown Source)
	at org.postgresql.ssl.LazyKeyManager.getCertificateChain(LazyKeyManager.java:159)
	... 60 more
2023-08-10 23:20:23,452 DEBUG middleware.log :: PUT /api/database/8 400 752.1 ms (6 DB calls) 
{:message "Could not open SSL certificate file /tmp/metabase-secret_9099912732785039104.tmp."}


### Information about your Metabase installation

```JSON
v46.6.4, but probably comes from before
```


### Severity

P2

### Additional context

The only workaround is to delete the DB and start from scratch or add a new DB and switch all questions to the new one

--- a/src/metabase/models/secret.clj
+++ b/src/metabase/models/secret.clj
@@ -143,6 +143,12 @@
   "Regex for parsing base64 encoded file uploads."
   #"^data:application/([^;]*);base64,")
 
+(defn latest-for-id
+  "Returns the latest Secret instance for the given `id` (meaning the one with the highest `version`)."
+  {:added "0.42.0"}
+  [id]
+  (t2/select-one Secret :id id {:order-by [[:version :desc]]}))
+
 (defn db-details-prop->secret-map
   "Returns a map containing `:value` and `:source` for the given `conn-prop-nm`. `conn-prop-nm` is expected to be the
   name of a connection property having `:type` `:secret`, and the relevant sub-properties (ex: -value, -path, etc.) will
@@ -185,14 +191,14 @@
                              {:invalid-db-details-entry (select-keys details [path-kw])}))))
 
                  (id-kw details)
-                 (:value (t2/select-one Secret :id (id-kw details))))
+                 (:value (latest-for-id (id-kw details))))
         source (cond
                  ;; set the :source due to the -path suffix (see above))
                  (and (not= "uploaded" (options-kw details)) (path-kw details))
                  :file-path
 
                  (id-kw details)
-                 (:source (t2/select-one Secret :id (id-kw details))))]
+                 (:source (latest-for-id (id-kw details))))]
     (cond-> {:connection-property-name conn-prop-nm, :subprops [path-kw value-kw id-kw]}
       value
       (assoc :value value
@@ -209,7 +215,7 @@
                       value
                       (String. ^bytes value "UTF-8")))
                   (when id
-                    (String. ^bytes (:value (t2/select-one Secret :id id)) "UTF-8")))]
+                    (String. ^bytes (:value (latest-for-id id)) "UTF-8")))]
     (case (options-kw details)
       "uploaded" (try
                    ;; When a secret is updated, the value has already been decoded
@@ -225,12 +231,6 @@
   ^{:doc "The attributes of a secret which, if changed, will result in a version bump" :private true}
   bump-version-keys
   [:kind :source :value])
-
-(defn latest-for-id
-  "Returns the latest Secret instance for the given `id` (meaning the one with the highest `version`)."
-  {:added "0.42.0"}
-  [id]
-  (t2/select-one Secret :id id {:order-by [[:version :desc]]}))
 
 (defn upsert-secret-value!
   "Inserts a new secret value, or updates an existing one, for the given parameters.
@@ -305,7 +305,7 @@
   (let [subprop (fn [prop-nm]
                   (keyword (str conn-prop-nm prop-nm)))
         secret* (cond (int? secret-or-id)
-                      (t2/select-one Secret :id secret-or-id)
+                      (latest-for-id secret-or-id)
 
                       (mi/instance-of? Secret secret-or-id)
                       secret-or-id

--- a/test/metabase/models/secret_test.clj
+++ b/test/metabase/models/secret_test.clj
@@ -210,3 +210,44 @@
               (is (bytes? decoded))
               (is (= content
                      (String. ^bytes decoded "UTF-8"))))))))))
+
+(deftest use-latest-version-test
+  (testing "when reading a secret from the DB, the latest version is taken (#33116)"
+    (let [initial-value "vnetillo"
+          latest-value (str initial-value "s")
+          initial-source "source"
+          latest-source (str initial-source "s")
+          secret-property "secret"
+          secret-map1 {:name  secret-property
+                       :kind  ::secret/password
+                       :value initial-value
+                       :source initial-source}
+          secret-map2 {:name  secret-property
+                       :kind  ::secret/password
+                       :value latest-value
+                       :source latest-source}
+          crowberto-id (mt/user->id :crowberto)
+          by-crowberto #(assoc % :creator_id crowberto-id)
+          timestamp-string? (fn [s]
+                              (try
+                                (java.time.Instant/parse s)
+                                true
+                                (catch Exception _
+                                  false)))]
+      (t2.with-temp/with-temp [Secret {:keys [id version]} (by-crowberto secret-map1)
+                               Secret _ (-> secret-map2
+                                            by-crowberto
+                                            (assoc :id id :version (inc version)))]
+        (let [details {:secret-id id}]
+          (testing "db-details-prop->secret-map"
+            (let [secret (secret/db-details-prop->secret-map details secret-property)]
+              (is (= latest-value (-> secret :value bytes (String. "UTF8"))))
+              (is (= (keyword latest-source) (:source secret)))))
+          (testing "get-secret-string"
+            (is (= latest-value (secret/get-secret-string details secret-property))))
+          (testing "expand-inferred-secret-values"
+            (is (=? {:secret-id id
+                     :secret-source (keyword latest-source)
+                     :secret-creator-id crowberto-id
+                     :secret-created-at timestamp-string?}
+                    (secret/expand-inferred-secret-values details secret-property id)))))))))

